hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceManager.java

```diff
@@ -653,6 +658,30 @@ public void regionServerRemoved(String regionserver) {
     transferQueues(regionserver);
   }
 
+  /**
+   * We want to run the peer removal in a separate thread when the peer removal
+   * is called from ReplicationSource shipper thread on encountering NoNodeException
+   * because peerRemoved terminate the source which might leave replication source
+   * in orphaned state.
+   * See HBASE-25741.
+   * @param peerId peer ID to be removed.
+   */
+  public void schedulePeerRemoval(final String peerId) {
+    final ExecutorService threadPool =
+      Executors.newSingleThreadExecutor(Threads.getNamedThreadFactory("peer-removal"));
+    try {
+      Future<?> res = threadPool.submit(new Runnable() {
+        @Override public void run() {
+          peerRemoved(peerId);
+        }
+      });
+      res.get();
```

Comment:
wait.. I don't get this.. the blocking thread here is the shipper thread, it should be killed before get() returns? How is this any different from the current state of code.

Reply:
@bharathv We are creating a new executor in this method and replication shipper thread calls this method and wait till it complete the execution. As part of this execution the shipper thread will also be terminated and this new thread make sure that it do the clean up after termination as well. In short, this make sure that thread calling the `peerRemoved` is not the one getting terminated within this method. Does it make sense?

Reply:
> replication shipper thread calls this method and wait till it complete the execution.

The thread is killed during wait and this thread pool is never cleaned up. My point is we don't need to wait on it. Instead submit this task to the "executor" in this class, that executor is shutdown eventually.